raiden/network/proxies/token.py

```diff
@@ -67,6 +146,44 @@ def allowance(
             )
         )
 
+    def approve_data(self, failed_at_blockhash: BlockSpecification) -> ApproveConditionsData:
+        balance = self.balance_of(self.client.address, failed_at_blockhash)
+        return ApproveConditionsData(balance=balance)
+
+    @contextmanager
+    def approve_transaction(
+        self, allowed_address: Address, allowance: TokenAmount
+    ) -> Iterator[ApprovePendingTransaction]:
+        """Minimal critical section to serialize allowance transactions.
+
+        Each channel deposit requires an allowance at least equal to the
+        deposit amount, because allowance itself is an idempontent operation
+        and by default the Raiden's proxies allow only the exact token amount
+        necessary for the deposit, an interesting race condition with
+        concurrent deposits. If the transactions are order as such:
+
+        - Approve1 + Approve2 + Deposit1 + Deposit2
+
+        One of the approvals will be a no-op and the second Deposit will fail.
+        To prevent the above error the following order must be enforced:
+
+        - Approve1 + Deposit1 + Approve2 + Deposit2
+
+        This context manager allows user code to forbid another approve
+        transaction from being sent concurrently, allowing the token network
+        proxy to send the first deposit transaction before the second approval
+        is sent.
+
+        This is necessary since the high-level `approve` method has a very
+        large critical section, which includes the mining and confirmation of
+        the transaction.
+        """
+        with self.token_lock:
+            unchecked = ApproveUncheckedTransaction(allowed_address, allowance)
+            pending = unchecked.estimate_gas(self)
+
+            yield pending
```

Comment:
At first I thought about forcing the transaction to be sent, but it is actually nice to allow the caller to give up, since there are additional checks performed by the other proxies (e.g. if there is a race that closes the channel, this allows the caller to not send both the approve and the deposit)